src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,645 @@
+
+namespace Microsoft.CodeAnalysis.CSharp.ConvertLinq
+{
+    internal sealed class CSharpConvertLinqQueryToLinqMethodProvider : AbstractConvertLinqProvider
+    {
+        private static readonly TypeSyntax VarNameIdentifier = SyntaxFactory.IdentifierName("var");
+
+        protected override IAnalyzer CreateAnalyzer(SemanticModel semanticModel, CancellationToken cancellationToken)
+            => new CSharpAnalyzer(semanticModel, cancellationToken);
+
+        private sealed class CSharpAnalyzer : AnalyzerBase<QueryExpressionSyntax, SyntaxNode, SyntaxNode>
+        {
+            public CSharpAnalyzer(SemanticModel semanticModel, CancellationToken cancellationToken)
+                : base(semanticModel, cancellationToken)
+            {
+            }
+
+            protected override string Title => CSharpFeaturesResources.Convert_linq_query_to_foreach;
+
+            protected override bool TryConvert(QueryExpressionSyntax source, out DocumentUpdate documentUpdate)
+            {
+                if (_semanticModel.GetDiagnostics(source.Span, _cancellationToken).Any())
+                {
+                    documentUpdate = default;
+                    return false;
+                }
+
+                if (source.DescendantTrivia().Any(trivia => trivia.MatchesKind(
+                    SyntaxKind.SingleLineCommentTrivia,
+                    SyntaxKind.MultiLineCommentTrivia,
+                    SyntaxKind.MultiLineDocumentationCommentTrivia) ||
+                    source.ContainsDirectives))
+                {
+                    documentUpdate = default;
+                    return false;
+                }
+
+                // TODO add a semantic check
+                // validate that all clauses are translated into methods with signatures corresponding to code generated by the refactoring.
+
+                switch (source.Parent.Kind())
+                {
+                    case SyntaxKind.ReturnStatement:
+                        return TryConvertInReturnStatement(source, out documentUpdate);
+                    case SyntaxKind.ForEachStatement:
+                        return TryConvertInForEach(source, out documentUpdate);
+                    case SyntaxKind.EqualsValueClause:
+                    case SyntaxKind.SimpleAssignmentExpression:
+                    case SyntaxKind.MultiplyAssignmentExpression:
+                        return TryReplaceWithLocalFunction(source, out documentUpdate);
+                    case SyntaxKind.ParenthesizedExpression:
+                        return TryConvertInParenthesizedExpression(source, out documentUpdate);
+                    default:
+                        documentUpdate = default;
+                        return false;
+                }
+            }
+
+            private bool TryConvertInParenthesizedExpression(QueryExpressionSyntax source, out DocumentUpdate documentUpdate)
+            {
+                // TODO any other cases?
```

Comment:
it seems like you shoudl probably do your checks here and then call back up to 'TryConvert'.  Consider, for example: ```return (from ...)```.  It looks like with your current logic, you wouldn't support this.